keras_cv/models/object_detection/yolo_v8/yolo_v8_label_encoder.py

```diff
@@ -0,0 +1,287 @@
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""Label encoder for YOLOV8. This uses the TOOD Task Aligned Assigner approach,
+and is adapted from https://github.com/ultralytics/ultralytics/blob/main/ultralytics/yolo/utils/tal.py
+"""  # noqa: E501
+
+import tensorflow as tf
+from tensorflow import keras
+from tensorflow.keras import layers
+
+from keras_cv.models.object_detection.yolo_v8.yolo_v8_iou_loss import bbox_iou
+
+
+def select_highest_overlaps(mask_pos, overlaps, max_num_boxes):
+    """Picks the anchor with the highest IoU with a GT box to break ties
+    when two GT boxes match to the same anchor"""
+    # (b, max_num_boxes, num_anchors) -> (b, num_anchors)
+    fg_mask = tf.reduce_sum(mask_pos, axis=-2)
+
+    def handle_anchor_with_two_gt_boxes(
+        fg_mask, mask_pos, overlaps, max_num_boxes
+    ):
+        mask_multi_gts = tf.repeat(
+            tf.expand_dims(fg_mask, axis=1) > 1, max_num_boxes, axis=1
+        )  # (b, max_num_boxes, num_anchors)
+        max_overlaps_idx = tf.argmax(overlaps, axis=1)  # (b, num_anchors)
+        is_max_overlaps = tf.one_hot(
+            max_overlaps_idx, max_num_boxes
+        )  # (b, num_anchors, max_num_boxes)
+        is_max_overlaps = tf.cast(
+            tf.transpose(is_max_overlaps, perm=(0, 2, 1)), overlaps.dtype
+        )  # (b, max_num_boxes, num_anchors)
+        mask_pos = tf.where(
+            mask_multi_gts, is_max_overlaps, mask_pos
+        )  # (b, max_num_boxes, num_anchors)
+        fg_mask = tf.reduce_sum(mask_pos, axis=-2)
+        return fg_mask, mask_pos
+
+    fg_mask, mask_pos = tf.cond(
+        tf.reduce_max(fg_mask) > 1,
+        lambda: handle_anchor_with_two_gt_boxes(
+            fg_mask, mask_pos, overlaps, max_num_boxes
+        ),
+        lambda: (fg_mask, mask_pos),
+    )
+
+    target_gt_idx = tf.argmax(mask_pos, axis=-2)  # (b, num_anchors)
+    return target_gt_idx, fg_mask, mask_pos
+
+
+def select_candidates_in_gts(xy_centers, gt_bboxes, epsilon=1e-9):
+    """Selects candidate anchors for GT boxes"""
+    n_anchors = xy_centers.shape[0]
+    bs, n_boxes, _ = gt_bboxes.shape
+
+    lt, rb = tf.split(
+        tf.reshape(gt_bboxes, (-1, 1, 4)), 2, axis=-1
+    )  # left-top, right-bottom
+    bbox_deltas = tf.reshape(
+        tf.concat(
+            [xy_centers[tf.newaxis] - lt, rb - xy_centers[tf.newaxis]], axis=2
+        ),
+        (-1, n_boxes, n_anchors, 4),
+    )
+
+    return tf.reduce_min(bbox_deltas, axis=-1) > epsilon
+
+
+@keras.utils.register_keras_serializable(package="keras_cv")
+class YOLOV8LabelEncoder(layers.Layer):
```

Comment:
You spent a ton of time understanding, validating, and rewriting this component.

Would you be able to provide a little knowledge diffusion in the docstring here?  You did a ton of work to understand this, and it would be great to have our docs from that!

Reply:
I've expanded the docstrings a lot. LMK what you think!

Reply:
Great work @ianstenbit!